### DISCUSSION TITLE
Prebuild performance/config/reset feedback

### DISCUSSION THREAD

**Discussion by repi** [community/community#13158] [Codespaces]
Thanks for opening up the Codespaces Prebuild support for enterprise users! 

Been experimenting with it for one of our larger Rust projects. And have some initial feedback and questions about them:

1. Would it be possible to configure / reduce the frequency of the prebuilds? 
    - Right now it will rebuild on every commit, as long as a prebuild is not already building, but we have quite few users of our Codespaces and as the prebuild is only used when someone is creating a new Codespace, it is for us rather wasteful to build it on every commit. 
    - Could one do configure to run the prebuild every day or week instead?

3. Is there a way to configure what type of machine runs the prebuild config? 
    - I assume there isn't as it it uses the current GitHub Action machines which are only 2 vCPU. On such machines our minimal prebuild takes 30 min, and a better prebuild for us (rust clippy, tests, release build) takes around 180 min which is quite significant and slow to iterate on. 
    - 8-16 vCPU machines would be preferred and would be happy to pay for it. 
    - But assume this will have to wait until GitHub Action in general supports more machine variations. Or is it possible for the prebuild to run on one's own self-hosted GitHub Action Runners?

5. It would be nice to have a way to "reset" an existing Codespace to the default, both in general without prebuilds but esp. when there are prebuilds. 
    - Now one have to destroy it and create a new one which feels like there should be a shortcut for, similar to how there is a shortcut in VS Code to "Rebuild container"
    - This is useful both for easy testing of the prebuild config, but also if one has a Codespace that is older and instead of manually syncing and building the latest in it, it would be faster and be able to just "reset" it to the latest available prebuild and go from there. Think this would increase the value of prebuilds quite significantly, esp. for larger projects that take quite a bit of time to build even from a pull.

**Comment by tanmayeekamath** [comment]
Hey @repi! Thanks so much for the detailed feedback. 

> Would it be possible to configure / reduce the frequency of the prebuilds?

Yes, we're working on adding support for scheduled prebuild configurations (this should be available early April). With that you will be able to configure the frequency and triggers to update your prebuild configuration. We will support 3 options
- Prebuild on every push (default)
- Prebuild on configuration change (only when your devcontainer is updated)
- Prebuild on a schedule (you can pick the days and time)

Would love any feedback on this. 

> Is there a way to configure what type of machine runs the prebuild config?

This isn't supported today but we're actively thinking about it. We have plans to add support for self hosted runners at which point you will have more control on the type of machine you want to use. And as you pointed out correctly, once Actions adds support for custom runners, we aim to also support those for prebuilding your codespaces. 
Curious what would work best for you - support for self hosted runners or custom runners or higher CPU GitHub hosted runners for prebuilding ? 

> It would be nice to have a way to "reset" an existing Codespace to the default, both in general without prebuilds but esp. when there are prebuilds.

We've typically seen codespaces to be ephemeral in nature -  a rebuild would also take similar amount of time as create to get the latest changes. That being said, with scheduled prebuilds you could have developers create a Codespace every week and set your prebuilds to update every week as well. The primary goal with prebuilds is to reduce longer creations times and a reset may end up being similar to a recreate from a fresh prebuild version. Please correct me if I misunderstood the need. 

Separately we're also working on making it easy to test changes made to a devcontainer on a repo and branch that has prebuilds enabled. We currently have the [following guidance in docs](https://docs.github.com/en/codespaces/prebuilding-your-codespaces/testing-dev-container-changes) to help ensure changes made to your configuration don't introduce breaking changes for your team. Would love to hear any feedback you  have on this aspect as well. 

Happy to jump on a call too to discuss this further if you'd like, please let me know and thanks for all the feedback again :)

**Comment by sinorrman** [comment]
Hi!

Now that larger Github-hosted runners are in Beta, I want to follow up on:
> Is there a way to configure what type of machine runs the prebuild config?

Is it currently possible to configure prebuilds with these runners? I have a repo that simply cannot build with less than 8GB of RAM, and would very much appreciate being able to use the larger runners for my prebuilds.

**Reply by edulix** [reply]
I came here looking for the same thing: please allow configurable runners for prebuilds. We have a project that takes more than 1h in current prebuilds.

**Comment by iambhabha** [comment]
Hi!

I'm experiencing a similar need for configurable runners for prebuilds. In our repository, the current prebuilds are taking significant time, and the ability to use larger runners with more than 8GB of RAM would greatly enhance our prebuild process.

If there's any information or updates regarding the configuration of runners for prebuilds, or if there are plans to support this in the future, I would greatly appreciate the details.

Thank you!
kapil

**Reply by edulix** [reply]
Just for the record, I think for now the best way to configure this using Github Actions, which you can configure to use [large runners](https://docs.github.com/en/actions/using-github-hosted-runners/about-larger-runners/about-larger-runners#machine-sizes-for-larger-runners). Basically, the idea is to use an image for the devcontainer that you have built using Github Actions. More info here:
• https://www.youtube.com/watch?v=M21loGvplVM&ab_channel=VisualStudioCode
• https://containers.dev/guide/prebuild (linked by the video)

**Comment by petersonjdNIH** [comment]
Adding a related feature request with a concrete proposal here: https://github.com/orgs/community/discussions/197411

Specifically, adding a **Quick Rebuild from prebuilt snapshot** option to the Rebuild Container dialog — shown only when a current prebuild exists for the active branch — so you get the same near-instant start as create-from-prebuild without having to delete and recreate the codespace.

Also addresses the earlier note in this thread that "a rebuild would take similar time as a create" — that hasn't been the case in practice. Rebuild runs the full build pipeline cold; only create pulls from the prebuilt snapshot. That gap is the core of the ask.
